src/cast-dispatch.c

```diff
@@ -81,6 +96,32 @@ static SEXP vec_cast_dispatch2(SEXP x, SEXP to,
     default:
       break;
     }
+    break;
+
+  case vctrs_class_bare_tibble:
+    switch (x_type) {
+    case vctrs_type_dataframe:
+      return df_as_dataframe(x, to, x_arg, to_arg);
+
+    case vctrs_type_s3:
+      switch (class_type(x)) {
+      case vctrs_class_bare_data_frame:
+        Rf_errorcall(R_NilValue, "`x` should have been classified as a `vctrs_type_dataframe`");
```

Comment:
Can you lead with `"Internal error:"` please? This will make it easier to search and replace all internal errors when we switch to a special error thrower/constructor for those.